### PR TITLE
Bugfix/2013 unused path parameter in paths

### PR DESCRIPTION
**Pull request by ralphdoe**
Add an error about unused path parameter definitions in paths #2013

### Description

As an editor, I want to know when path parameters are not properly defined.

Swagger Editor currently has a validation for path templates that are missing path parameter definitions (e.g. /foo/{bar} without the bar parameter defined). It would be great to also check for the opposite situation - when a path parameter is defined but not included in the actual path string.

Definitions containing path parameters that are not included in the path string are invalid according to the OpenAPI Specification:

If in is "path", the name field MUST correspond to the associated path segment from the path field in the Paths Object.

### Motivation and Context

https://github.com/swagger-api/swagger-editor/issues/2013

### How Has This Been Tested?

`path parameters must be in path definition` section added in validate-semantic/2and3/parameters.js

4 Test cases added, and all previous tests are working properly.

### My PR contains... 

- [ ] No code changes (`src/` is unmodified: changes to documentation, CI, metadata, etc.)
- [ ] Dependency changes (any modification to dependencies in `package.json`)
- [ ] Bug fixes (non-breaking change which fixes an issue)
- [x] Improvements (misc. changes to existing features)
- [ ] Features (non-breaking change which adds functionality)

### My changes...
- [ ] are breaking changes to a public API (config options, System API, major UI change, etc).
- [ ] are breaking changes to a private API (Redux, component props, utility functions, etc.).
- [ ] are breaking changes to a developer API (npm script behavior changes, new dev system dependencies, etc).
- [x] are not breaking changes.

### Documentation
- [x] My changes do not require a change to the project documentation.
- [ ] My changes require a change to the project documentation.
- [ ] If yes to above: I have updated the documentation accordingly.

### Automated tests
- [ ] My changes can not or do not need to be tested.
- [x] My changes can and should be tested by unit and/or integration tests.
- [x] If yes to above: I have added tests to cover my changes.
- [x] If yes to above: I have taken care to cover edge cases in my tests.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/plugins/validate-semantic/validators/2and3/parameters.js [hunk 1]
```diff
@@ -73,3 +73,48 @@ export const validate2And3ParameterDefaultsMatchAnEnum = () => (system) => {
...
+  return system.validateSelectors
+    .allParameters()
+    .then(nodes => {
+      return nodes.reduce((acc, node) => {
+        const parameter = node.node || {}
+        const path = node.path
+        const isFromPath = path[0] === "paths" ? true : false
+        const pathString = path[1]
+        const paramName = parameter.name
+        const paramInPath = "{" +  paramName + "}"
```
Comment (tim-lai) [inline src/plugins/validate-semantic/validators/2and3/parameters.js:88]: prefer template literal. e.g. `{${paramName}}`

File: src/plugins/validate-semantic/validators/2and3/parameters.js [hunk 2]
```diff
@@ -73,3 +73,48 @@ export const validate2And3ParameterDefaultsMatchAnEnum = () => (system) => {
...
+        const ref = parameter.$ref
+        if (parameter !== undefined) {
+          if (parameter.in === "path") {
```
Comment (tim-lai) [inline src/plugins/validate-semantic/validators/2and3/parameters.js:91]: let's try to flatten these if/else. since you are already defining some `const` to check the conditionals, let's use more.

```
const parameterInPath = parameter.in === "path" ? true: false
const pathStringIncludesParamInPath = ...
if (parameterInPath && isFromPath && pathStringIncludesParamInPath) {
acc.push(...)
}
```

repeat below for other nested if/else

File: src/plugins/validate-semantic/validators/2and3/parameters.js [hunk 3]
```diff
@@ -73,3 +73,48 @@ export const validate2And3ParameterDefaultsMatchAnEnum = () => (system) => {
...
+            if (isFromPath) {
+              if (pathString !== undefined && !pathString.toUpperCase().includes("" + paramInPath.toUpperCase())) {
```
Comment (tim-lai) [inline src/plugins/validate-semantic/validators/2and3/parameters.js:93]: you could use `if (pathString && ...`, which would also check for `null`

repeat below

File: src/plugins/validate-semantic/validators/2and3/parameters.js [hunk 4]
```diff
@@ -73,3 +73,48 @@ export const validate2And3ParameterDefaultsMatchAnEnum = () => (system) => {
...
+    .then(nodes => {
+      return nodes.reduce((acc, node) => {
+        const parameter = node.node || {}
+        const path = node.path
+        const isFromPath = path[0] === "paths" ? true : false
+        const pathString = path[1]
+        const paramName = parameter.name
+        const paramInPath = "{" +  paramName + "}"
+        const ref = parameter.$ref
+        if (parameter !== undefined) {
```
Comment (tim-lai) [inline src/plugins/validate-semantic/validators/2and3/parameters.js:90]: this `undefined` check seems extraneous, since you assign an initial value in line 83 above

Comment (ralphdoe) [reply src/plugins/validate-semantic/validators/2and3/parameters.js:90]: Done! Thanks!

File: src/plugins/validate-semantic/validators/2and3/parameters.js (hunk 3 above)
Comment (ralphdoe) [reply src/plugins/validate-semantic/validators/2and3/parameters.js:93]: Done! Cool!

File: src/plugins/validate-semantic/validators/2and3/parameters.js (hunk 2 above)
Comment (ralphdoe) [reply src/plugins/validate-semantic/validators/2and3/parameters.js:91]: Done!

File: src/plugins/validate-semantic/validators/2and3/parameters.js (hunk 1 above)
Comment (ralphdoe) [reply src/plugins/validate-semantic/validators/2and3/parameters.js:88]: Done!
